internal/cephfs/core/quiesce.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2024 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+
+	"github.com/ceph/ceph-csi/internal/util"
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	"github.com/ceph/go-ceph/cephfs/admin"
+)
+
+type QuiesceStates string
+
+const (
+	Released  QuiesceStates = "RELEASED"
+	Quiescing QuiesceStates = "QUIESCING"
+	Quiesced  QuiesceStates = "QUIESCED"
+)
+
+func GetQuiesceState(set *admin.QuiesceInfo) QuiesceStates {
+	state := QuiesceStates(set.State.Name)
+	switch set.State.Name {
+	case "RELEASED":
+		state = Released
+	case "QUIESCING":
+		state = Quiescing
+	case "QUIESCED":
+		state = Quiesced
+	}
+
+	return state
+}
+
+type SubVolumeQuiesceClient interface {
+	SubvolumeQuiesce(
+		ctx context.Context,
+		reserveName string,
+	) (*admin.QuiesceInfo, error)
+	GetVolumes() []Volume
+	SubvolumeQuiesceWithExpireTimeout(ctx context.Context,
+		reserveName string,
+	) (*admin.QuiesceInfo, error)
+	ResetSubvolumeQuiesce(ctx context.Context,
+		reserveName string,
+	) (*admin.QuiesceInfo, error)
+	ReleaseSubvolumeQuiesce(ctx context.Context,
+		reserveName string,
+	) (*admin.QuiesceInfo, error)
+}
+
+type Volume struct {
+	VolumeID  string
+	ClusterID string
+}
+
+type subVolumeQuiesce struct {
+	fsName                string
+	volumes               []Volume
+	subVolumeGroupMapping map[string][]string
+	fsa                   *admin.FSAdmin
+}
+
+func NewSubVolumeQuiesce(
+	fsName string,
+	volumes []Volume,
+	mapping map[string][]string,
+	conn *util.ClusterConnection,
+) (SubVolumeQuiesceClient, error) {
+	fsa, err := conn.GetFSAdmin()
+	if err != nil {
+		return nil, err
+	}
+
+	return &subVolumeQuiesce{
+		fsName:                fsName,
+		volumes:               volumes,
+		subVolumeGroupMapping: mapping,
+		fsa:                   fsa,
+	}, nil
+}
+
+func (svq *subVolumeQuiesce) GetVolumes() []Volume {
+	return svq.volumes
+}
+
+func (svq *subVolumeQuiesce) getSubvolumeNamesString() []string {
```

Comment:
Please explain in a comment what this function does. What is the original contents look like, and how are the `volName` (plural?)` constructed?

Reply:
updated function name to getMembers(), `ceph fs quiesce` CLI calls this as members and added a comment for it with example